### PR TITLE
testsuite: a test case for a pessimization from 13076

### PR DESCRIPTION
**Pull request by gasche**
#13076, part of The Pattern-Matching Bug PR series and already merged, introduced a small regression / pessimization -- some pattern-matching programs generate slightly worse code than before. I will be working on a fix, but a good first step is to introduce a testcase for the issue -- which would help reviewing a fix, and can serve as a regression test later.

### CODE REVIEW

**Comment by ncik-roberts** [comment]
(I happen to have been reading through the linked PR and saw this, hence the immediate response.) This indeed looks like a proper regression test. I would click "Approve" if I could.

**Comment by gasche** [comment]
I added the two extra test cases proposed by @ncik-roberts in https://github.com/ocaml/ocaml/pull/13076#issuecomment-2256960227 .

**Review by ncik-roberts — APPROVED** [review]

**Comment by gasche** [comment]
This needs approval from a maintainer to move forward.

**Review by yallop — APPROVED** [review]
Approving on behalf of @ncik-roberts (but it looks sensible to me, too).

**Comment by gasche** [comment]
Thanks! Merging.

**Pull request merged**
